Fix inductor torch.cat edge case for empty tensor

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107193

Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @chillee @jansel @eellison

wconstab committed:
Fix inductor torch.cat edge case for empty tensor

Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc @chillee @jansel @eellison

[ghstack-poisoned]

wanchaol commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -122,7 +122,15 @@ def normalize_cat_default(match: Match, *args, **kwargs):
+        x_shape = x.meta["example_value"].shape
+        return len(x_shape) == 1 and x_shape[0] == 0
+
+    assert all(
```
curious in general for special case like this, do we filter out empty tensors in the fx pass directly (i.e. remove that node from `tensors`), or we do this when creating inductor IR?

wconstab replied:
good q. i only intended to do the change to ir.py, but the test i added failed on the pass.  i am not sure if there is a better way to handle this.

wanchaol reviewed (approved):
lgtm, one question

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2549,6 +2549,15 @@ def create(cls, inputs, dim):
+        # just drop the 'empty' inputs so they don't confuse the logic below.
+        def non_empty_input(input):
+            input_size = input.get_size()
+            return len(input.get_size()) > 1 or input.get_size()[0] > 0
```
right now the assumption is that if a size is 0, it will always be guarded on.. i dont know how much we should be thinking about unbacked symints in the future.

wconstab replied:
which decomp would that be, there is a decomp for cat?  if so how come we are lowering cat into inductor?

eellison replied:
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py#L209-L212. It just exists to simplify one case.


Could extend the decomp so that if there are no 0-dim inputs, return NotImplemented. Otherwise, call cat recursively with those elements removed.

wconstab replied:
ok- I pushed a change to move the special case out of ir.py and into decomps.  However, I couldn't delete the special case from split_cat since iiuc its a pre-autograd pass.  Please lmk if this version is good to land

eellison reviewed (approved):
looks good, you could consider putting this logic in the decomp: 

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py#L208-L212

peterbell10 commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -122,7 +122,15 @@ def normalize_cat_default(match: Match, *args, **kwargs):
         return
 
     ndim = tensors[0].meta["example_value"].dim()
```
This line isn't valid since empty tensors don't need to have the same number of dimensions as the rest of the arguments. Can we take ndim from the output tensor instead?

wconstab replied:
hm, perhaps `cat_node.meta['example_value'].dim()`?  i'll also rewrite the warning to check cat_node and all the tensors for 'example_value'

wconstab committed:
Update on "Fix inductor torch.cat edge case for empty tensor"


Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov chillee jansel eellison

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/decomposition.py:
```diff
@@ -207,6 +207,13 @@ def mm(self, input2):
+        # just drop the 'empty' inputs so they don't confuse the logic below.
+        return len(x.shape) > 1 or x.shape[0] > 0
+
+    tensors = list(filter(non_empty_tensor, tensors))
```
This doesn't do anything, you need to call `aten.cat` if the lists has changed, i.e.:
```
if len(filtered_tensors) != len(tensors):
    return aten.cat(filtered_tensors, dim=dim)
```

And you'll probably need a special case for the empty list.

wconstab replied:
ahh you're right. but why does my test pass :/

wconstab committed:
Update on "Fix inductor torch.cat edge case for empty tensor"


Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov chillee jansel eellison

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/decomposition.py:
```diff
@@ -207,8 +207,19 @@ def mm(self, input2):
+
+    if len(filtered_tensors) == 1:
         return tensors[0].clone()
+    elif len(filtered_tensors) > 1 and len(filtered_tensors) < len(tensors):
```
NIT: 
```suggestion
    elif 1 < len(filtered_tensors) < len(tensors):
```

wconstab committed:
Update on "Fix inductor torch.cat edge case for empty tensor"


Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov chillee jansel eellison

[ghstack-poisoned]

wconstab committed:
Update on "Fix inductor torch.cat edge case for empty tensor"


Align with eager behavior on this edge case- essentially, the empty
tensor is ignored by the operator.

Fixes #107118
cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov chillee jansel eellison

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge

ezyang commented on torch/_inductor/decomposition.py:
```diff
@@ -212,8 +212,19 @@ def mm(self, input2):
-    if len(tensors) == 1:
+    def non_empty_tensor(x):
+        # special case for cat'ing with an empty tensor -
+        # just drop the 'empty' inputs so they don't confuse the logic below.
```
I'm trying to understand in what sense you want to avoid confusing the logic below; and my best guess is that you're actually referring to torch/_inductor/fx_passes/split_cat.py (where you are enforcing an invariant that there are no malformed inputs to cat.) Is this correct?

peterbell10 replied:
I think the comment is refering to the lowering for `aten.cat` which doesn't have the special case for empty 1d tensors. In eager tensors with shape `[0]` are ignored completely by the cat operator, and don't need to have the correct shape in the non-dim dimensions or even the correct number of dimensions.

ezyang replied:
Hmm, but this seems questionable, we should be on the hook for accurately lowering aten.cat as it is defined. If we wanted a simplified version of it, we ought to have a prim... which we do! It's called prims.cat.
